traitsui/tests/editors/test_set_editor.py

```diff
@@ -0,0 +1,592 @@
+
+            # Button inserts at the end
+            self.assertEqual(
+                get_list_items(editor._unused), ["four", "three", "one", "two"]
+            )
+            self.assertEqual(get_list_items(editor._used), [])
+
+    @skip_if_null
+    def test_simple_set_editor_default_selection_unused(self):
+        with store_exceptions_on_all_threads():
+            gui, editor = self.setup_gui(ListModel(), get_view())
+
+            self.assertEqual(get_list_items(editor._unused), ["four", "three"])
+            self.assertEqual(get_list_items(editor._used), ["one", "two"])
+
+            click_button(editor._use)
+            gui.process_events()
+
+            # Button inserts at the top
+            self.assertEqual(
+                get_list_items(editor._unused), ["three"]
+            )
+            self.assertEqual(
+                get_list_items(editor._used), ["four", "one", "two"]
+            )
+
+    @skip_if_null
+    def test_simple_set_editor_default_selection_used(self):
+        # When all items are used, top used item is selected by default
+        list_edit = ListModel(value=["one", "two", "three", "four"])
+
+        with store_exceptions_on_all_threads():
+            gui, editor = self.setup_gui(list_edit, get_view())
+
+            self.assertEqual(get_list_items(editor._unused), [])
+            self.assertEqual(
+                get_list_items(editor._used), ["four", "one", "three", "two"])
+
+            click_button(editor._unuse)
+            gui.process_events()
+
+            # Button inserts at the top
+            self.assertEqual(get_list_items(editor._unused), ["four"])
+            self.assertEqual(
+                get_list_items(editor._used), ["one", "three", "two"]
+            )
+
+    @skip_if_null
+    def test_simple_set_editor_deleted_valid_values(self):
+        editor_factory = SetEditor(values=["one", "two", "three", "four"])
+        view = View(UItem("value", editor=editor_factory, style="simple",))
+        list_edit = ListModel()
+
+        with store_exceptions_on_all_threads():
+            gui, editor = self.setup_gui(list_edit, view)
+
+            self.assertEqual(get_list_items(editor._unused), ["four", "three"])
+            self.assertEqual(get_list_items(editor._used), ["one", "two"])
+
+            editor_factory.values = ["two", "three", "four"]
+            gui.process_events()
+
+            self.assertEqual(get_list_items(editor._unused), ["four", "three"])
+            if is_current_backend_wx():  # FIXME issue #840
+                with self.assertRaises(AssertionError):
+                    self.assertEqual(get_list_items(editor._used), ["two"])
+                self.assertEqual(get_list_items(editor._used), ["one", "two"])
+            else:
+                self.assertEqual(get_list_items(editor._used), ["two"])
+            self.assertEqual(list_edit.value, ["two"])
+
+    @skip_if_null
+    def test_simple_set_editor_use_ordered_selected(self):
```

Comment:
This test looks almost identical to `test_simple_set_editor_use_button`.
It seems that the only difference is the last assertion: where the moved item is selected again if `ordered` is true. Did I understand correctly?

Reply:
The setup is a also bit different because of `ordered` flag. But yes, apart from that the tests are identical, although I don't think they should necessarily be. If the editor has `ordered=False`, `update_editor` sorts the values alphabetically, but it isn't called on button clicks right now.

Reply:
Hmmm maybe I am still confused what `ordered` is designed to do.

It looks like the most important effect of the `ordered=True` flag is:
- Show the list of values using the order as is given on the trait
- Show the move-up and move-down arrows for manually reordering values on the trait.

This is tied to whether the values being edited (in this case: `ListModel.value`) is ordered or not. Currently `ListModel.value` is `["one", "two"]`, which happens to be alphabetically ordered.

If the `ListModel.value` is changed to `["two", "one"]` with `ordered=True`, `update_editor` will be called but it won't try to alphabetically sort the values, then we expect the view to show `["two", "one"]`, as opposed to `["one", "two"]` (alphabetically ordered).

Could we add this check or does it exist here already? It could be done by instantiating `ListModel(value=["two", "one"])` instead of `ListModel()` and check the used list is `["two", "one"]`.

(The fact that the selection is clear when `ordered=False` looks to me a minor UX optimization, but I am only guessing on this point.)

Reply:
I expanded both tests to show the sorting behaviour

Reply:
Thank you!